### DISCUSSION TITLE
Local DUAs and exceptions for fault localization

### DISCUSSION THREAD

**Discussion by Alex-128** [saeg/ba-dua#39] [Q&A] [answered]
Hi again!

I've continued working with BA-DUA for my project and I'm very happy with it, but I did realize there are two issues, namely that DUAs within one basic block are not tracked, and that DUAs that trigger an exception aren't marked as covered. 

I did some research before coming here, and from the description in your original paper on BA-DUA (An efficient bitwise algorithm for intra-procedural data-flow testing coverage), it seems that only DUAs with a path from the def basic block and the use basic block are tracked. I'm curious as to why this is so? Was this purely a design choice, or was there another reason? I feel that tracking DUAs in the same basic block would add more information, such as in the worked example in your paper, where the faulty data-flow is the def and use of `i` in basic block 1. It also doesn't feel like it would be too difficult to implement, since such a DUA would be covered as soon as its definition line is executed. I was wondering whether you have any fixes for this, but since the Jaguar paper notes it as a limitation, I would guess not. If you have some advice for adjusting BA-DUA for this purpose (I assume during the instrumentation process), I would very much appreciate it.

The other issue, which the Jaguar paper also mentions, is exceptions. This feels like a more tricky issue to solve, but I would also appreciate advice if you have any.

Kind regards
Alexandra

**Comment by andrioli** [comment, marked as answer]
Okay let me try to respond these tricky questions.

Let me start with the easiest ones: local DUAs.

> namely that DUAs within one basic block are not tracked

This is true!

> it seems that only DUAs with a path from the def basic block and the use basic block are tracked. I'm curious as to why this is so? Was this purely a design choice, or was there another reason?

This was by design. But why?  Because we don't need to track each individual DUA and because local DUA a trivial to track.

>  I feel that tracking DUAs in the same basic block would add more information, such as in the worked example in your paper, where the faulty data-flow is the def and use of i in basic block 1.

Definitely true

> It also doesn't feel like it would be too difficult to implement, since such a DUA would be covered as soon as its definition line is executed.

This is basically why I said they are trivial to track. Assuming that all instructions of a basic block are always executed given the first instructions is reach, to track local DUAs we just need to track if the basic block was reach, reducing the problem to control-flow coverage. But due to exception flow, this assumptions could not be held. 

We are also studying DUAs subsumption. A global DUA can subsumes a local DUA. So no need to track each individual local DUA.

> I was wondering whether you have any fixes for this, but since the Jaguar paper notes it as a limitation, I would guess not. If you have some advice for adjusting BA-DUA for this purpose (I assume during the instrumentation process), I would very much appreciate it.

We didn't implemented support for local DUAs. As I said before, this was a design choice, but maybe, in the future, we will include this on BA-DUA.

If you want to implement it, my suggestions is to track which basic blocks were covered, and then on report phase, detect witch local DUAs were covered  by simply verifying local DUAs within a basic block.

As always, contributions are welcome, I'll be happy to review this if you open a merge request to BA-DUA repository.

Regarding exceptions, they are more tricky, I'll need some time to elaborate a more complete response, but we already support them somehow (in an optimistic way). In the meanwhile you can review BA-DUA's validation tests on exceptions: [[1]](https://github.com/saeg/ba-dua/blob/ba-dua-0.8.0/ba-dua-tests/src/test/java/br/usp/each/saeg/badua/test/validation/CatchExceptionSourceTest.java), [[2]](https://github.com/saeg/ba-dua/blob/ba-dua-0.8.0/ba-dua-tests/src/test/java/br/usp/each/saeg/badua/test/validation/NotCatchExceptionSourceTest.java), [[3]](https://github.com/saeg/ba-dua/blob/ba-dua-0.8.0/ba-dua-tests/src/test/java/br/usp/each/saeg/badua/test/validation/MaxException1SourceTest.java) and [[4]](https://github.com/saeg/ba-dua/blob/ba-dua-0.8.0/ba-dua-tests/src/test/java/br/usp/each/saeg/badua/test/validation/MaxException2SourceTest.java).

**Reply by Alex-128** [reply]
Thanks for the reply! I will look into this.

**Comment by andrioli** [comment]
I did a POC  a long time ago to support local duas (https://github.com/saeg/ba-dua/tree/local-duas). But I was not happy with the code, so I didn't integrate it to mainline.

**Reply by Alex-128** [reply]
Hi, I hope you've been well! I have been using this code to track local DUAs as well, and it's worked great for me so far. However, when I was recently testing JFreeChart fault localization, bug 26 performed very poorly since the faulty DUA (1191, 1192, owner) was assigned a 0.0 suspiciousness score. This is despite the fact that the exception occurs within the same basic block as the (1191, 1192, owner) DUA and the optimistic exception support you indicated above (and which is clearly supported in the local DUA code). I was wondering whether you have had this problem before, or if it's an issue on my side?

Kind regards
Alexandra

**Reply by andrioli** [reply]
This is probably due to the exceptions flow problem. 

Can you execute this bug with the following JVM flag `-Dbadua.experimental.exception_handler=true` to verify it fix the problem ?

TL;DR

I owe you an answer on exceptions, so here goes:

The problem with exception flows is that they occur abruptly, at least from BA-DUA current static analysis perspective.  So for BA-DUA, exception are not expected. After all, they are exceptions 😄.

The exception flows complicates static analysis and control flow graph too much, and that was the reason that we choose to not fully support it on BA-DUA. There is an old tool (JaBuTi) that do support exception flows, I think there are some papers that explain how they handle it. 

But how we handle exceptions ?

First you need to know that BA-DUA monitor the DUAs on an optimistic way. This means that once a basic block is reached during runtime execution, we assume that all instructions on that basic block were executed. And not a surprise that this is not true when exceptions occurs.

Second, we keep track of covered DUAs locally (don't get confused by local DUAs, we only keep track if were covered or not as local variables in the current method) and this informations is made available to BA-DUA later, just before the method exits. By exits I mean that execution reached a `return` statement (or an implicit return) or a `throw` statement.

What happens when a method exits abruptly by an exception flow ? In this case, historically, the coverage is lost. But this changed when we introduced a catch and thrown approach. Basically, during the method execution if any unexpected exception occurs, we catch it, made the coverage information available to BA-DUA, and then re-thrown the same exception.

See:  commit 5ec4f340158f8c062e8fb94bbd1b999dcfbf85b9

This feature was introduced disable by default

See: commit 3c967c7b2f21f2a133e273587a04d60f0144866e

And it was integrated to mainline on commit 9b483f12c62aa6d6159c46774a3bbb8e015d2ae8 and released on version 0.4.0.

Finally, on commit 2ba7e34cea96978ec4f2cc47a6bf7326dceaeb91 we enabled this feature by default and released on version 0.5.0.

The support for local DUA is based on the version 0.4.0, but as I said before, was never integrated in the mainline, and it was made before we enabled the catch and thrown by default. So you need to enable it manually if using this version.

Please, also note that the catch and thrown feature introduced a regression for some edge case methods that was only fixed on pull request #31 and released only in version 0.7.0

I hope it helps!

**Reply by Alex-128** [reply]
Thanks for the quick reply, as always! I do remember you telling me about the optimistic exceptions, which is why I was so surprised by Chart 26 not having the elements covered for the failing test case. I wasn't sure where the JVM flag, so I've added it to the instrumentation (`java -Dbadua.experimental.exception_handler=true -jar ba-dua-cli-0.81-SNAPSHOT-all.jar instrument -dest instrumented -src build/`), the execution, and the reporting, if that's what you meant? It doesn't seem to change the ranking :(

**Reply by andrioli** [reply]
If using the offline instrumentation the flag should be added to the `instrument` command. You used the correct command.

Looking to the command looks that you used BA-DUA version 0.8. Recall that this version does not have support for local DUAs. Unless you integrated the branch [local-duas](https://github.com/saeg/ba-dua/tree/local-duas) to the 0.8 version.

Can you share with the me the instrumented classes for `source/org/jfree/chart/axis/Axis.java`. I think this is the class that have the local DUA, right ? So I can do a manual inspection in the generated bytecode.

**Reply by Alex-128** [reply]
Yes, I did integrate the branch into the 0.8 version and it does show the local DUAs correctly and measure their coverage optimistically (I tried the `max` method in your _Evaluating data-flow coverage in spectrum-based fault localization_ paper with array {1} and it does mark the DUA from the def of `i` to its use in defining `max` as suspicious).

Here's my instrumented file for Chart 26:
[Axis_instr.zip](https://github.com/saeg/ba-dua/files/14773974/Axis_instr.zip)
(sorry for the zip file, but my GitHub complains about uploading a class file)

**Reply by andrioli** [reply]
Thanks, 

After looking the provided bytecode and the source code of the local-duas branch, I found a few problems on the implementation.

As I said before, I didn't integrate local-duas branch to mainline because I was not happy with the code. Besides the code complex and difficult to understand, looks that we have bugs.

For methods that the sum of DUAs and basic blocks are higher than 64 entities, the current implementation will not work! The code that does the initialization should be fixed [here](https://github.com/saeg/ba-dua/blob/8d9402c569085729941449ccff694babb7fe56ca/ba-dua-core/src/main/java/br/usp/each/saeg/badua/core/internal/instr/CoverageMethodTransformer.java#L152-L159) and [here](https://github.com/saeg/ba-dua/blob/8d9402c569085729941449ccff694babb7fe56ca/ba-dua-core/src/main/java/br/usp/each/saeg/badua/core/internal/instr/CoverageMethodTransformer.java#L238-L246).

Anyway, I think the support for local duas will be a great addition to BA-DUA. So I'll spend some time trying to fix it.

I keep you posted here!

**Reply by andrioli** [reply]
And note that, since you integrated local-duas branch on version 0.8 in your local repository, you probably don't need the option `-Dbadua.experimental.exception_handler=true` as this feature is enabled by default on latest versions.

**Reply by andrioli** [reply]
@Alex-128 I did a quick workaround on the local-duas branch. See commit https://github.com/saeg/ba-dua/commit/5ae1abacc9fcd25cfc76f79f699317227d180176.

I didn't tested with the Chart bug 26, but at least the BA-DUA unit tests are still green.

Can you test if those changes will fix the problem to you ?

If so, I'll start to design a definitive implementation to support local DUAs.

**Reply by Alex-128** [reply]
Thank you, I just tested and it does work for Chart 26! Interestingly enough, it ranks the faulty DUA (1191,1192,owner), which triggers the NullPointerException, as more suspicious than the DUAs that have the "faulty" line 1192 as def. I'm unsure if that was intentional on your side, but whatever works :) Thanks again for all your effort, you have helped me so much!

**Reply by andrioli** [reply]
I'm glad it worked.

>Interestingly enough, it ranks the faulty DUA (1191,1192,owner), which triggers the NullPointerException, as more suspicious than the DUAs that have the "faulty" line 1192 as def. I'm unsure if that was intentional on your side

I think this is the correct behavior, since the bug occurs in the use of the variable with value null. 

next weeks I'll try to integrate this feature to the mainline definitely. I'll appreciate any feedback.

**Reply by Alex-128** [reply]
I would be happy to give feedback. Thank you for all the help, once again!

**Comment by andrioli** [comment]
So, just to be clear, since BA-DUA monitor the DUAs on an optimistic way, in case of exceptions, some DUAs that were not covered may be marked as covered.
